File: 1742. Maximum Number of Balls in a Box/Solution.py
class Solution:
    def countBalls(self, lowLimit: int, highLimit: int) -> int:
        #return max(self.bruteforce(lowLimit, highLimit))
        return max(self.pattern(lowLimit, highLimit))

    def bruteforce(self, lowLimit, highLimit):
        counts = [0 for i in range(46)]
        for i in range(lowLimit, highLimit + 1):
            currSum = 0
            while i > 0:
                currSum += i % 10
                i //= 10
            counts[currSum] += 1
        return counts

    def pattern(self, lowLimit, highLimit):
        counts = [0 for i in range(46)]
        start, pointer = lowLimit, 0
        while start > 0:
            pointer += start % 10
            start //= 10
        counts[pointer] += 1
        for i in range(lowLimit + 1, highLimit + 1):
            while i % 10 == 0:
                pointer -= 9
                i //= 10
            pointer += 1
            counts[pointer] += 1
        return counts

File: 1742. Maximum Number of Balls in a Box/Solution.java
class Solution {
    public int countBalls(int lowLimit, int highLimit) {
        // int[] counts = bruteforce(lowLimit, highLimit);
        int[] counts = pattern(lowLimit, highLimit);
        int res = 0;
        for (int c : counts) res = Math.max(res, c);
        return res;
    }

    public int[] bruteforce(int lowLimit, int highLimit) {
        int[] counts = new int[46];
        for (int i = lowLimit; i < highLimit + 1; i++) {
            int curr = i, currSum = 0;
            while (curr > 0) {
                currSum += curr % 10;
                curr /= 10;
            }
            counts[currSum]++;
        }
        return counts;
    }

    public int[] pattern(int lowLimit, int highLimit) {
        int[] counts = new int[46];
        int low = lowLimit, pointer = 0;
        while (low > 0) {
            pointer += low % 10;
            low /= 10;
        }
        counts[pointer]++;
        for (int i = lowLimit + 1; i < highLimit + 1; i++) {
            int curr = i;
            while (curr % 10 == 0) {
                pointer -= 9;
                curr /= 10;
            }
            pointer++;
            counts[pointer]++;
        }
        return counts;
    }
}
